### PR TITLE
Progressbar on docker pull

### PR DESCRIPTION
**Pull request by Williangalvani**
https://user-images.githubusercontent.com/4013804/196457525-f8a8fa54-2056-40c5-9be4-befa993377ca.mov

### CODE REVIEW

**Review by rafaellehmkuhl — APPROVED** [review]

File: core/frontend/src/utils/pull_tracker.ts [hunk 1]
```diff
@@ -42,6 +93,9 @@ class PullTracker {
             if ('status' in data) {
               this.layer_status[id] = data.status
             }
+            if ('progressDetail' in data && 'total' in data.progressDetail) {
```
Comment (patrickelectric) [inline core/frontend/src/utils/pull_tracker.ts:96]:
```suggestion
            if (data.?progressDetail.?total !== undefined) {
```

File: core/frontend/src/utils/pull_tracker.ts [hunk 2]
```diff
@@ -15,10 +20,56 @@ class PullTracker {
...
+          extraction_total += this.layer_progress_detail[id].total
+        }
+        if (this.layer_status[id] === 'Extracting') {
+          // if we are extracting, we finished downloading
+          download_total += this.layer_progress_detail[id].total
+          download_current += this.layer_progress_detail[id].total
+
+          extraction_total += this.layer_progress_detail[id].total
+          extraction_current += this.layer_progress_detail[id].current
+        }
```
Comment (patrickelectric) [inline core/frontend/src/utils/pull_tracker.ts:67]: The if cascade can be organized to a switch/case.
```suggestion
        switch (this.layer_status[id]) {
            case 'Waiting':
                download_total += this.guesstimated_size
                extraction_total += this.guesstimated_size
                break
                
            case 'Downloading':
                download_total += this.layer_progress_detail[id].total
                download_current += this.layer_progress_detail[id].current
                extraction_total += this.layer_progress_detail[id].total
                break
                
            case 'Pull complete':
                download_total += this.layer_progress_detail[id].total
                download_current += this.layer_progress_detail[id].total
                extraction_total += this.layer_progress_detail[id].total
                extraction_current += this.layer_progress_detail[id].total
                break
                
            case 'Download complete':
                download_total += this.layer_progress_detail[id].total
                download_current += this.layer_progress_detail[id].total
                extraction_total += this.layer_progress_detail[id].total
                break
                
            case 'Extracting':
                download_total += this.layer_progress_detail[id].total
                download_current += this.layer_progress_detail[id].total
                extraction_total += this.layer_progress_detail[id].total
                extraction_current += this.layer_progress_detail[id].current
                break
        }
```

File: core/frontend/src/utils/pull_tracker.ts [hunk 3]
```diff
@@ -3,10 +3,15 @@ import { Dictionary } from '@/types/common'
...
     private layer_status: Dictionary<string> = {}
 
+    // We need to assume a size for the missing layers, we are assuming 100mb for now.
+    private guesstimated_size = 100 * 1024 * 1024
+
     private layers: string[] = []
 
     private layer_progress: Dictionary<string> = {}
 
+    private layer_progress_detail: Dictionary<Dictionary<number>> = {}
```
Comment (patrickelectric) [inline core/frontend/src/utils/pull_tracker.ts:13]: Isn't this a `Dictionary<{total: number, current: number}> = {}` ?

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/frontend/src/utils/pull_tracker.ts (hunk 3 above)
Comment (Williangalvani) [reply core/frontend/src/utils/pull_tracker.ts]: I didn't remember you could do that. nice.

File: core/frontend/src/views/ExtensionManagerView.vue [hunk 4]
```diff
@@ -192,15 +198,24 @@ export default Vue.extend({
...
         .catch((error) => {
           notifier.pushBackError('EXTENSIONS_INSTALL_FAIL', error)
         })
+        .finally(() => {
+          this.show_dialog = false
+          this.pull_output = ''
+          this.download_percentage = 0
+          this.extraction_percentage = 0
+          this.status_text = ''
+        })
```
Comment (joaoantoniocardoso) [inline core/frontend/src/views/ExtensionManagerView.vue:218]: note here that we are not going back to the initial state, as `download_total` and `extraction_total` are not being zeroed.

Comment (Williangalvani) [reply core/frontend/src/views/ExtensionManagerView.vue:218]: these are not available in this context. plus we use a new instance of pulltracker for each pull, so things in there don't require re-zeroing

**Review by patrickelectric — APPROVED** [review]

**Pull request merged**